// goose.js/src/ConfigProperties.js
import _ from 'lodash';
import readYaml from 'read-yaml';
import YAML from 'json2yaml';
import Path from 'path';
import FileUtils from './utils/FileUtils';
import Out from './utils/Out';

const jsonConfigTemplate = {
  environments: {
    default_migration_table: 'goose_migrations',
    default_database: 'development',
    development: {
      adapter: 'mysql',
      host: 'localhost',
      // name: 'production_db',
      user: 'root',
      pass: '',
      port: 3306,
      charset: 'utf8',
      database: '',
      // collation: 'utf8_unicode_ci',
    },
  },
  paths: {
    migrations: 'db/migrations',
    seeds: 'db/seeds',
  },
};

export { jsonConfigTemplate };
export default class Config {
  constructor(params) {
    this.homeDir = _.get(params, 'homeDir', '.');

    this.environments = _.get(params, 'environments', {});
    this.environments.default_migration_table = _.get(params,
      'environments.default_migration_table',
      jsonConfigTemplate.environments.default_migration_table);

    this.environments.default_database = _.get(params,
      'environments.default_database');

    this.paths = {
      migrations: _.get(params, 'paths.migrations', jsonConfigTemplate.paths.migrations),
      seeds: _.get(params, 'paths.seeds', jsonConfigTemplate.paths.seeds),
    };
    this.environment = null;
  }
  get templateDir() {
    return Path.join(__dirname, '../template');
  }
  get defaultMigrationTable() {
    return _.get(this.environments, 'default_migration_table', null);
  }
  get defaultDatabase() {
    return _.get(this.environments, 'default_database', null);
  }
  getEnvironmentDatabase(envName) {
    return _.get(this.environments, envName, null);
  }
  get database() {
    return this.getEnvironmentDatabase(this.environment);
  }
  get migrationsDir() {
    return Path.join(this.homeDir, this.paths.migrations);
  }
  get seedsDir() {
    return Path.join(this.homeDir, this.paths.seeds);
  }

  static templateConfig() {
    return JSON.stringify(jsonConfigTemplate, null, 2);
  }
  static templateConfigYAML() {
    return YAML.stringify(jsonConfigTemplate);
  }

  static async from(filePath) {
    const data = await this.readFile(filePath);
    return new Config(data);
  }
  static async readFile(filePath) {
    if (!FileUtils.exists(filePath)) {
      Out.warn(`Can't resolve config file: ${filePath}`);
      return null;
    }

    let parsedObj = null;

    if (/\.json$/.test(filePath)) {
      // read json file
      parsedObj = JSON.parse(FileUtils.read(filePath));
    } else if (/\.js$/.test(filePath)) {
      // read json file
      parsedObj = require(filePath);  // eslint-disable-line
    } else if (/\.yml$/.test(filePath)) {
      // const text = fs.readFileSync(filePath, 'utf8');
      parsedObj = readYaml.sync(filePath);
    } else {
      Out.warn(`Bad config file format, only json|yml are supported, found: ${filePath}`);
    }
    return parsedObj;
  }
}
