// goose.js/dist/DatabaseHandler.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _knex = require('knex');

var _knex2 = _interopRequireDefault(_knex);

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

var _Error = require('./Error');

var _Logger = require('./utils/Logger');

var _Logger2 = _interopRequireDefault(_Logger);

var _FileUtils = require('./utils/FileUtils');

var _FileUtils2 = _interopRequireDefault(_FileUtils);

var _Helpers = require('./utils/Helpers');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var log = (0, _Logger2.default)(__filename);

/**
 * Returns provider connection settings.
 * @param provider
 * @param params
 * @returns {*}
 */
var getConnecionConfig = function getConnecionConfig(adapter, params) {
  var map = {
    mysql: {
      client: 'mysql',
      version: '5.6',
      connection: {
        host: _lodash2.default.get(params, 'host'),
        port: _lodash2.default.get(params, 'port'),
        user: _lodash2.default.get(params, 'user'),
        password: _lodash2.default.get(params, 'pass'),
        database: _lodash2.default.get(params, 'database')
      }
    },
    pgsql: {
      client: 'pg',
      connection: {
        host: _lodash2.default.get(params, 'host'),
        port: _lodash2.default.get(params, 'port'),
        user: _lodash2.default.get(params, 'user'),
        password: _lodash2.default.get(params, 'pass'),
        database: _lodash2.default.get(params, 'database')
      },
      searchPath: _lodash2.default.get(params, 'searchPath', ['public'])
    },
    sqlite: {
      client: 'sqlite',
      connection: {
        filename: _lodash2.default.get(params, 'filename')
      }
    }
  };
  return map[adapter];
};

var Handler = function () {
  function Handler() {
    _classCallCheck(this, Handler);
  }

  _createClass(Handler, [{
    key: 'initializeTable',
    value: async function initializeTable(tableNameOverride) {
      var tableName = !tableNameOverride ? this.mTable : tableNameOverride;
      var exists = await this.tableExists(tableName);
      if (exists) {
        log.debug('Table already exists ' + tableName);
        return null;
      }
      return this.createMigrationTable(tableName);
    }
  }, {
    key: 'tableExists',


    /**
     *
     * @param tableName
     * @returns {Promise<boolean>}
     */
    value: async function tableExists(tableName) {
      var bool = false;
      try {
        bool = await this.connection.hasTable(tableName);
      } catch (e) {
        //
      }
      return bool;
    }

    /**
     * Drop table if exists.
     * @param table
     * @returns {Promise<void>}
     */

  }, {
    key: 'dropTableIfExists',
    value: async function dropTableIfExists(table) {
      return this.connection.dropTableIfExists(table);
    }

    /**
     * Create migration table
     * @param tableName
     * @returns {Promise<void>}
     */

  }, {
    key: 'createMigrationTable',
    value: async function createMigrationTable(tableName) {
      return this.connection.createTable(tableName, function (table) {
        table.string('id').notNullable();
        table.string('name').notNullable();
        table.string('status').notNullable();
        table.timestamp('start_time').notNullable();
        table.timestamp('end_time').notNullable();
        table.timestamp('created_at').notNullable();
      });
    }

    /**
     * Close the connection
     */

  }, {
    key: 'close',
    value: async function close() {
      return this.knex.destroy();
    }

    /**
     * Select all migrations
     * @param table
     * @returns {Promise<void>}
     */

  }, {
    key: 'allFiles',
    value: async function allFiles() {
      return this.knex.orderBy('id', 'asc').orderBy('name', 'asc').orderBy('created_at', 'asc').from(this.mTable);
    }

    /**
     * Select all migrations
     * @param table
     * @returns {Promise<void>}
     */

  }, {
    key: 'mergedFiles',
    value: async function mergedFiles() {
      return this.knex.from(this.mTable).orderBy('id', 'asc').orderBy('name', 'asc').orderBy('created_at', 'asc').where({ status: 'up' });
    }

    /**
     * Execute a migration file
     * @param table
     * @returns {Promise<void>}
     */

  }, {
    key: 'truncate',
    value: async function truncate(tableName) {
      return this.knex(tableName).truncate();
    }

    /**
     * Execute a migration file
     * @param table
     * @returns {Promise<void>}
     */

  }, {
    key: 'exec',
    value: async function exec(filePath) {
      var SQL = _FileUtils2.default.read(filePath);
      log.debug('SQL file content: ' + SQL);

      var lines = Handler.splitStatements(SQL);
      var startTime = (0, _Helpers.isoFormat)(new Date());
      var _iteratorNormalCompletion = true;
      var _didIteratorError = false;
      var _iteratorError = undefined;

      try {
        for (var _iterator = lines[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
          var line = _step.value;
          // eslint-disable-line
          await this.connection.raw(line);
        }
      } catch (err) {
        _didIteratorError = true;
        _iteratorError = err;
      } finally {
        try {
          if (!_iteratorNormalCompletion && _iterator.return) {
            _iterator.return();
          }
        } finally {
          if (_didIteratorError) {
            throw _iteratorError;
          }
        }
      }

      var endTime = (0, _Helpers.isoFormat)(new Date());
      return Promise.resolve({
        startTime: startTime,
        endTime: endTime,
        lines: lines.length
      });
    }

    /**
     * Merge migration.
     * @param id
     * @param filePath
     * @param name
     * @returns {Promise<*>}
     */

  }, {
    key: 'merge',
    value: async function merge(id, filePath, name) {
      var _ref = await this.exec(filePath),
          startTime = _ref.startTime,
          endTime = _ref.endTime;

      var UP = 'up';

      if (await this.migrationIDExists(id)) {
        return this.knex(this.mTable).where({ id: id }).update({
          status: UP,
          start_time: startTime,
          end_time: endTime
        });
      }

      return this.knex(this.mTable).insert({
        id: id,
        name: name,
        status: UP,
        start_time: startTime,
        end_time: endTime,
        created_at: endTime
      });
    }

    /**
     * Revert
     * @param id
     * @param filePath
     * @returns {Promise<*>}
     */

  }, {
    key: 'revert',
    value: async function revert(id, filePath) {
      var _ref2 = await this.exec(filePath),
          startTime = _ref2.startTime,
          endTime = _ref2.endTime;

      return this.knex(this.mTable).update({
        status: 'down',
        start_time: startTime,
        end_time: endTime
      }).where({ id: id });
    }

    /**
     * Check if migration exists.
     * @param id
     * @returns {Promise<boolean>}
     */

  }, {
    key: 'migrationIDExists',
    value: async function migrationIDExists(id) {
      var row = await this.knex(this.mTable).where({ id: id }).select().first();
      return row !== undefined;
    }

    /**
     *
     * @param context
     * @returns {*}
     */

  }, {
    key: 'connection',
    get: function get() {
      return this.knex.schema;
    }
  }], [{
    key: 'create',

    /**
     * Create instance
     * @param provider
     * @param params
     */
    value: async function create(params) {
      var adapter = params.adapter;

      if (!Handler.validAdapter(adapter)) {
        throw new _Error.DBInvalidProvider('Wrong database provider: ' + adapter);
      }

      var instance = new Handler();
      instance.provider = adapter;
      instance.config = getConnecionConfig(adapter, params);
      instance.mTable = 'goose_migrations';
      instance.knex = (0, _knex2.default)(instance.config);
      await instance.initializeTable();
      return instance;
    }

    /**
     * Validate database provider
     * @param provider
     * @returns {boolean}
     */

  }, {
    key: 'validAdapter',
    value: function validAdapter(adapter) {
      return (/(mysql)|(sqlite)|(pgsql)/.test(('' + adapter).toLowerCase())
      );
    }
  }, {
    key: 'splitStatements',
    value: function splitStatements(text) {
      if (!text) {
        return [];
      }
      var lines = text.trim().split(';'); // eslint-disable-line
      return (0, _lodash2.default)(lines).map(function (it) {
        return it.trim();
      }).value(); // trim every line
    }
  }]);

  return Handler;
}();

exports.default = Handler;
